Code/Source/RobotImporter/URDF/VisualsMaker.cpp

```diff
@@ -0,0 +1,208 @@
+#include <LmbrCentral/Shape/EditorShapeComponentBus.h>
+#include <LmbrCentral/Shape/SphereShapeComponentBus.h>
+
+namespace ROS2
+{
+    VisualsMaker::VisualsMaker(const AZStd::string& modelPath, const std::map<std::string, urdf::MaterialSharedPtr>& materials)
+        : m_modelPath(modelPath)
+        , m_materials(materials)
+    {
+    }
+
+    void VisualsMaker::AddVisuals(urdf::LinkSharedPtr link, AZ::EntityId entityId)
+    {
+        const AZStd::string typeString = "visual";
+        if (link->visual_array.size() == 0)
+        { // one or zero visuals - element is used
+            AddVisual(link->visual, entityId, PrefabMakerUtils::MakeEntityName(link->name.c_str(), typeString));
+            return;
+        }
+
+        size_t nameSuffixIndex = 0; // For disambiguation when multiple unnamed visuals are present. The order does not matter here
+        for (auto visual : link->visual_array)
+        { // one or more visuals - array is used
+            AddVisual(visual, entityId, PrefabMakerUtils::MakeEntityName(link->name.c_str(), typeString, nameSuffixIndex));
+            nameSuffixIndex++;
+        }
+    }
+
+    void VisualsMaker::AddVisual(urdf::VisualSharedPtr visual, AZ::EntityId entityId, const AZStd::string& generatedName)
+    {
+        if (!visual)
+        { // it is ok not to have a visual in a link
+            return;
+        }
+
+        if (!visual->geometry)
+        { // non-empty visual should have a geometry
+            AZ_Warning("AddVisual", false, "No Geometry for a visual");
+            return;
+        }
+
+        AZ_TracePrintf("AddVisual", "Processing visual for entity id:%s", entityId.ToString().c_str());
+
+        // Use a name generated from the link unless specific name is defined for this visual
+        const char* subEntityName = visual->name.empty() ? generatedName.c_str() : visual->name.c_str();
+        // Since O3DE does not allow origin for visuals, we need to create a sub-entity and store visual there
+        auto createEntityResult = PrefabMakerUtils::CreateEntity(entityId, subEntityName);
+        if (!createEntityResult.IsSuccess())
+        {
+            AZ_Error("AddVisual", false, "Unable to create a sub-entity for visual element %s", subEntityName);
+            return;
+        }
+        auto visualEntityId = createEntityResult.GetValue();
+        AddVisualToEntity(visual, visualEntityId);
+        AddMaterialForVisual(visual, visualEntityId);
+    }
+
+    void VisualsMaker::AddVisualToEntity(urdf::VisualSharedPtr visual, AZ::EntityId entityId)
+    {
+        // Apply transform as per origin
+        PrefabMakerUtils::SetEntityTransform(visual->origin, entityId);
+
+        AZ::Entity* entity = AzToolsFramework::GetEntityById(entityId);
+        auto geometry = visual->geometry;
+        switch (geometry->type)
+        {
+        case urdf::Geometry::SPHERE:
+            {
+                auto sphereGeometry = std::dynamic_pointer_cast<urdf::Sphere>(geometry);
```

Comment:
RTTI is disabled by default on the global compile settings. Modules that need RTTI will need to explicitly enable it. Is this relying on RTTI? If so, there doesn't seem to be a nullptr check for this object

Reply:
I keep unresolved, I activated RTTI and added assertions. I think it is not need to operate, but the bug in URDF parser can cause undefined behavior here without RTTI. Please take a look @adamdbrw

Reply:
The use of RTTI is fine as long as this code is not client/game code.

Reply:
Actually we don't disable disable RTTI explicitly anymore.
We use the default value for the compiler(which is on by default for most compilers).